### PR TITLE
tsp, json-merge-patch, add `JsonMergePatchHelper`

### PR DESCRIPTION
**Pull request by haolingdong-msft**
Link https://github.com/Azure/autorest.java/issues/1530

Code Changes:
1. Add `JsonMergePatchTemplate.java` and related facotry code.
2. `TypeSpecPlugin.java`: generate JsonMergePatchHelper only when there is model used in json-merge-patch operation and needs to generate model.
3. `TemplateUtil.java`: add JsonMergePatchHelper class's name into it.

Generated `JsonMergePatchHelper` example: typespec-tests/src/main/java/com/_specs_/azure/core/basic/implementation/JsonMergePatchHelper.java

PS: Name the helper class `JsonMergePatchHelper` rather than `XXXHelper` (XXX is service name), because later we will need to use the helper in model, but `ModelTemplate` does not have the context of service name. Discussed with @weidongxu-microsoft  for simplicity, we will call it `JsonMergePatchHelper` for now.

image

### CODE REVIEW

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java [hunk 1]
```diff
@@ -1230,4 +1230,11 @@ private static void loadStringOrArraySettingAsArray(String settingName, Consumer
             action.accept(settingValues);
         }
     }
+
+    private final String jsonMergePatchHelperClassName = "JsonMergePatchHelper";
+
+    public String getJsonMergePatchHelperClassName() {
+        return jsonMergePatchHelperClassName;
+    }
```
Comment (weidongxu-microsoft) [inline extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java:1238]: just put it in some util class.

File: javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+import com.azure.autorest.model.javamodel.JavaClass;
+import com.azure.autorest.model.javamodel.JavaFile;
+import com.azure.autorest.model.javamodel.JavaVisibility;
+import com.azure.autorest.util.ClientModelUtil;
+import com.azure.autorest.util.CodeNamer;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class JsonMergePatchHelperTemplate implements IJavaTemplate<Client, JavaFile>{
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java:17]: Does it really need the whole `Client`?

If not, provide what is necessary, e.g. `List<ClientModel>`

File: javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    public static JsonMergePatchHelperTemplate getInstance() {
+        return INSTANCE;
+    }
+
+    @Override
+    public void write(Client client, JavaFile javaFile) {
+        // imports
+        JavaSettings settings = JavaSettings.getInstance();
+        Set<String> imports = new HashSet<>();
+        addImports(imports, client, settings);
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java:33]: same, why import on `Client`? Does the code has anything to do with `Client`?

File: typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java [hunk 4]
```diff
@@ -57,6 +58,15 @@ protected void writeClientModels(Client client, JavaPackage javaPackage, JavaSet
...
                 .forEach(model -> javaPackage.addModel(model.getPackage(), model.getName(), model));
 
+        // JsonMergePatchHelper
+        boolean hasJsonMergePatchModelToGenerate = client.getModels().stream()
+                .filter(ModelUtil::isGeneratingModel)
+                .filter(ClientModelUtil::isJsonMergePatchModel)
+                .findFirst().isPresent();
+        if (hasJsonMergePatchModelToGenerate) {
+            javaPackage.addJsonMergePatchHelper(client);
+        }
```
Comment (weidongxu-microsoft) [inline typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java:68]: Adjust this accordingly.

**Review by weidongxu-microsoft — COMMENTED** [review]
Before merge, make sure to turn off one link of the flow so no additional file will be generated till the whole thing is ready to work.

File: typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java [hunk 5]
```diff
@@ -57,6 +58,15 @@ protected void writeClientModels(Client client, JavaPackage javaPackage, JavaSet
...
                 .forEach(model -> javaPackage.addModel(model.getPackage(), model.getName(), model));
 
+        // JsonMergePatchHelper
+        boolean hasJsonMergePatchModelToGenerate = client.getModels().stream()
+                .filter(ModelUtil::isGeneratingModel)
+                .filter(ClientModelUtil::isJsonMergePatchModel)
+                .findFirst().isPresent();
+//        if (hasJsonMergePatchModelToGenerate) {
+//            javaPackage.addJsonMergePatchHelper(client.getModels());
+//        }
```
Comment (weidongxu-microsoft) [inline typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java:68]: recommend to filter the models (as you already done here), and only send these to the template.

and template no longer need to do the filter again.

File: javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+
+    /**
+     * Add getter and setter for each model that is used in json-merge-patch.
+     *
+     * @param models
+     * @param javaClass
+     */
+    private void addGettersAndSetters(List<ClientModel> models, JavaClass javaClass) {
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java:101]: use static method whenever possible.

File: javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        if (models != null && !models.isEmpty()) {
+            models.forEach(model -> {
+                if (ClientModelUtil.isJsonMergePatchModel(model)) {
+                    // setters
+                    javaClass.publicStaticMethod(String.format("void set%sAccessor(%sAccessor accessor)", model.getName(), model.getName()),methodBlock -> {
+                        methodBlock.line(String.format("%sAccessor = accessor;", CodeNamer.getModelNamer().modelPropertySetterName(model.getName())));
+                    });
+                    // getters
+                    javaClass.publicStaticMethod(String.format("%sAccessor get%sAccessor()", model.getName(), model.getName()), methodBlock -> {
+                        methodBlock.line(String.format("return %sAccessor;",CodeNamer.getModelNamer().modelPropertySetterName(model.getName())));
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java:111]: it should have a `methodReturn` method.

File: javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    private void addGettersAndSetters(List<ClientModel> models, JavaClass javaClass) {
+        if (models != null && !models.isEmpty()) {
+            models.forEach(model -> {
+                if (ClientModelUtil.isJsonMergePatchModel(model)) {
+                    // setters
+                    javaClass.publicStaticMethod(String.format("void set%sAccessor(%sAccessor accessor)", model.getName(), model.getName()),methodBlock -> {
+                        methodBlock.line(String.format("%sAccessor = accessor;", CodeNamer.getModelNamer().modelPropertySetterName(model.getName())));
+                    });
+                    // getters
+                    javaClass.publicStaticMethod(String.format("%sAccessor get%sAccessor()", model.getName(), model.getName()), methodBlock -> {
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/JsonMergePatchHelperTemplate.java:110]: Use e.g. `%1$s` whenever you need to use same string more than once in format.

File: typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java (hunk 5 above)
Comment (haolingdong-msft) [reply typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java:68]: sure, will pass in the filtered models to addJsonMergePatchHelper(), but 'm not sure if we want to delete the checking logic in template, as template should not assume all the passed-in models are json merge patch model, it's safer to add  the check in template as well? What do you think?

Comment (weidongxu-microsoft) [reply typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java:68]: Of course don't delete the check logic. It just becomes "if list is not empty"

Comment (haolingdong-msft) [reply typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java:68]: I mean  the logic on checking whether model is json merge patch model. Do we want to keep it for safety?

Comment (weidongxu-microsoft) [reply typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java:68]: No. For safety you should do null check on every single string usage, do you?

Comment (haolingdong-msft) [reply typespec-extension/src/main/java/com/azure/autorest/TypeSpecPlugin.java:68]: ok, fine.

File: javagen/src/main/java/com/azure/autorest/util/TemplateUtil.java [hunk 9]
```diff
@@ -59,6 +59,8 @@ public class TemplateUtil {
 
     public static final String DATE_UTC = "date-utc";
 
+    public static final String JSON_MERGE_PATCH_HELPER_CLASS_NAME = "JsonMergePatchHelper";
+
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/TemplateUtil.java:63]: move to `TemplateHelper` or `ClientModelUtil`

at present `TemplateUtil` is for the replacement of placeholder strings.

File: javagen/src/main/java/com/azure/autorest/model/javamodel/JavaPackage.java [hunk 10]
```diff
@@ -321,4 +322,11 @@ protected void checkDuplicateFile(String filePath) {
             logger.warn(String.format("Name conflict for output file '%1$s'.", filePath));
         }
     }
+
+    public void addJsonMergePatchHelper(List<ClientModel> models) {
+        JavaFile javaFile = javaFileFactory.createSourceFile(settings.getPackage("implementation"), ClientModelUtil.JSON_MERGE_PATCH_HELPER_CLASS_NAME);
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/model/javamodel/JavaPackage.java:327]:
```suggestion
        JavaFile javaFile = javaFileFactory.createSourceFile(settings.getPackage(settings.getImplementationSubpackage()), ClientModelUtil.JSON_MERGE_PATCH_HELPER_CLASS_NAME);
```

**Review by weidongxu-microsoft — APPROVED** [review]

**Pull request merged**